--- ../original/MatMult_SeqAIJ.c	2017-08-14 15:28:14.302854000 -0400
+++ MatMult_SeqAIJ.c	2017-08-14 14:28:06.000000000 -0400
@@ -46,14 +46,42 @@
     aa   = a->a;
     fortranmultaij_(&m,x,ii,aj,aa,y);
 #else
+
+    // alias to CSR data
+    const MatScalar *data = a->a;
+    const PetscInt *cols = a->j;
+
+    // variable for size of x array
+    PetscInt xSize;
+
+    // get size of x array
+    ierr = VecGetLocalSize(xx, &xSize); CHKERRQ(ierr);
+
+    // upload matrix if it does not exist on device
+    # pragma acc enter data copyin(ii[:m+1]) wait
+    # pragma acc enter data copyin(cols[:a->nz])
+    # pragma acc enter data copyin(data[:a->nz])
+
+    // upload vector x
+    # pragma acc enter data copyin(x[:xSize])
+
+    // GPU kernels
+    # pragma acc kernels loop independent gang vector(32)\
+      present(ii[:m+1], cols[:a->nz], data[:a->nz], x[:xSize]) \
+      copyout(y[:m])
     for (i=0; i<m; i++) {
       n           = ii[i+1] - ii[i];
-      aj          = a->j + ii[i];
-      aa          = a->a + ii[i];
+      aj          = cols + ii[i];
+      aa          = data + ii[i];
       sum         = 0.0;
+      # pragma acc loop seq reduction(+:sum)
       PetscSparseDensePlusDot(sum,x,aa,aj,n);
       y[i] = sum;
     }
+
+    // delete vector x
+    # pragma acc exit data delete(x[:xSize])
+
 #endif
   }
   ierr = PetscLogFlops(2.0*a->nz - a->nonzerorowcnt);CHKERRQ(ierr);
