--- ../original/MatDestroy_SeqAIJ.c	2017-08-03 18:16:43.981950000 -0400
+++ MatDestroy_SeqAIJ.c	2017-07-31 14:16:56.816633000 -0400
@@ -5,2 +5,5 @@
 
+// OpenACC header
+# include <openacc.h>
+
 #undef __FUNCT__
@@ -16,2 +19,18 @@
 #endif
+
+  // OpenACC starts here
+  int present[3];
+
+  PetscInt *ai = a->i;
+  PetscInt *aj = a->j;
+  MatScalar *aa = a->a;
+
+  present[0] = acc_is_present(ai, (A->rmap->n+1)*sizeof(PetscInt));
+  present[1] = acc_is_present(aj, (a->nz)*sizeof(PetscInt));
+  present[2] = acc_is_present(aa, (a->nz)*sizeof(MatScalar));
+  # pragma acc exit data delete(ai[0:A->rmap->n+1]) if(present[0]) async(1)
+  # pragma acc exit data delete(aj[0:a->nz]) if(present[1]) async(1)
+  # pragma acc exit data delete(aa[0:a->nz]) if(present[2]) async(1)
+  // OpenACC ends here
+
   ierr = MatSeqXAIJFreeAIJ(A,&a->a,&a->j,&a->i);CHKERRQ(ierr);
